pkg/controller/certificatesigningrequest/ipsec_csr_signing_controller.go

```diff
@@ -0,0 +1,486 @@
+
+	selfSignedCA bool
+
+	// saved CertificateAuthority
+	certificateAuthority atomic.Value
+
+	queue         workqueue.RateLimitingInterface
+	fixturesQueue workqueue.RateLimitingInterface
+}
+
+// certificateAuthority implements a certificate authority and used by the signing controller.
+type certificateAuthority struct {
+	// RawCert is an optional field to determine if signing cert/key pairs have changed
+	RawCert []byte
+	// RawKey is an optional field to determine if signing cert/key pairs have changed
+	RawKey []byte
+
+	Certificate *x509.Certificate
+	PrivateKey  crypto.Signer
+}
+
+func (c *certificateAuthority) signCSR(template *x509.Certificate, requestKey crypto.PublicKey) (*x509.Certificate, error) {
+	if len(c.RawCert) == 0 || len(c.RawKey) == 0 {
+		return nil, fmt.Errorf("certificate authority is not valid")
+	}
+	derBytes, err := x509.CreateCertificate(rand.Reader, template, c.Certificate, requestKey, c.PrivateKey)
+	if err != nil {
+		return nil, err
+	}
+	certs, err := x509.ParseCertificates(derBytes)
+	if err != nil {
+		return nil, err
+	}
+	if len(certs) != 1 {
+		return nil, fmt.Errorf("expect a single certificate, got %d", len(certs))
+	}
+	return certs[0], nil
+}
+
+// NewIPsecCSRSigningController returns a new *IPsecCSRSigningController.
+func NewIPsecCSRSigningController(client clientset.Interface,
+	csrInformer csrinformer.CertificateSigningRequestInformer,
+	selfSignedCA bool,
+) (*IPsecCSRSigningController, error) {
+
+	caConfigMapInformer := corev1informers.NewFilteredConfigMapInformer(client, env.GetAntreaNamespace(), 12*time.Hour, cache.Indexers{cache.NamespaceIndex: cache.MetaNamespaceIndexFunc}, func(listOptions *v1.ListOptions) {
+		listOptions.FieldSelector = fields.OneTermEqualSelector("metadata.name", ipsecRootCAName).String()
+	})
+
+	configmapLister := corev1listers.NewConfigMapLister(caConfigMapInformer.GetIndexer())
+
+	c := &IPsecCSRSigningController{
+		client:                client,
+		csrLister:             csrInformer.Lister(),
+		csrListerSynced:       csrInformer.Informer().HasSynced,
+		configMapInformer:     caConfigMapInformer,
+		configMapLister:       configmapLister,
+		configMapListerSynced: caConfigMapInformer.HasSynced,
+		selfSignedCA:          selfSignedCA,
+		queue:                 workqueue.NewNamedRateLimitingQueue(workqueue.NewItemExponentialFailureRateLimiter(minRetryDelay, maxRetryDelay), "certificateSigningRequest"),
+		fixturesQueue:         workqueue.NewNamedRateLimitingQueue(workqueue.NewItemExponentialFailureRateLimiter(minRetryDelay, maxRetryDelay), "certificateSigningRequest"),
+	}
+
+	csrInformer.Informer().AddEventHandlerWithResyncPeriod(
+		cache.ResourceEventHandlerFuncs{
+			AddFunc: c.enqueueCertificateSigningRequest,
+			UpdateFunc: func(old, cur interface{}) {
+				c.enqueueCertificateSigningRequest(cur)
+			},
+			DeleteFunc: c.enqueueCertificateSigningRequest,
+		},
+		resyncPeriod,
+	)
+
+	caConfigMapInformer.AddEventHandlerWithResyncPeriod(
```

Comment:
Why it needs to watch caConfigMap? doesn't just need to create/update it once when CA cert changes?

Reply:
It's not a common mistake to delete a configMap, it requires specific permissions and exact command to do such operation. I think it normally means some programing error of some manifest management softwares, and recovering from such situation may hide the issue that the configMap is deleted unexpectedly (we did meet such issue before) and would perhaps lead to a loop (one software keeps deleting the configMap while antrea keeps recreating it). And anyway, it's still recoverable by just restarting the controller, which should make sense to user even if they deleted the map manually by mistake.

Reply:
After more thinking, I agree it would be better if it can recover automatically if the cost is not high. But syncRootCertificateAndKey should use the cache of caConfigMapInformer instead of calling Get API since it has watched it.

Reply:
ok. updated.